frontend/src/metabase/querying/parameters/utils/query.ts

```diff
@@ -0,0 +1,223 @@
+import { P, match } from "ts-pattern";
+
+import { getDateFilterClause } from "metabase/querying/filters/utils/dates";
+import * as Lib from "metabase-lib";
+import { isTemporalUnitParameter } from "metabase-lib/v1/parameters/utils/parameter-type";
+import type {
+  ParameterTarget,
+  ParameterType,
+  ParameterValueOrArray,
+  StructuredParameterDimensionTarget,
+} from "metabase-types/api";
+import { isStructuredDimensionTarget } from "metabase-types/guards";
+
+import {
+  normalizeBooleanParameterValue,
+  normalizeDateParameterValue,
+  normalizeNumberParameterValue,
+  normalizeStringParameterValue,
+} from "./normalize";
+
+const STRING_OPERATORS: Partial<
+  Record<ParameterType, Lib.StringFilterOperator>
+> = {
+  "string/=": "=",
+  "string/!=": "!=",
+  "string/contains": "contains",
+  "string/does-not-contain": "does-not-contain",
+  "string/starts-with": "starts-with",
+  "string/ends-with": "ends-with",
+};
+
+const NUMBER_OPERATORS: Partial<
+  Record<ParameterType, Lib.NumberFilterOperator>
+> = {
+  "string/=": "=",
+  "number/=": "=",
+  "number/!=": "!=",
+  "string/!=": "!=",
+  "number/>=": ">=",
+  "number/<=": "<=",
+  "number/between": "between",
+};
+
+export function applyParameter(
+  query: Lib.Query,
+  stageIndex: number,
+  type: ParameterType,
+  target: ParameterTarget | null,
+  value: ParameterValueOrArray | null,
+) {
+  if (target == null || value == null || !isStructuredDimensionTarget(target)) {
+    return query;
+  }
+
+  if (isTemporalUnitParameter(type)) {
+    return applyTemporalUnitParameter(query, stageIndex, target, value);
+  } else {
+    return applyFilterParameter(query, stageIndex, type, target, value);
+  }
+}
+
+function applyFilterParameter(
+  query: Lib.Query,
+  stageIndex: number,
+  type: ParameterType,
+  target: StructuredParameterDimensionTarget,
+  value: ParameterValueOrArray,
+): Lib.Query {
+  const column = getParameterFilterColumn(query, stageIndex, target);
+  if (column == null) {
+    return query;
+  }
+
+  const filter = getParameterFilterClause(type, column, value);
+  if (filter == null) {
+    return query;
+  }
+
+  return Lib.filter(query, stageIndex, filter);
+}
+
+function getParameterFilterColumn(
```

Comment:
This function should also become shared in the future but I don't understand what the API should be yet; leaving it local.